### PR TITLE
use getReportName and getParentNavigationSubtitle from ReportUtils for reports with title or subtitle as empty string for QR code page

### PR DESCRIPTION
**Pull request by c3024**
### Fixed Issues

$ https://github.com/Expensify/App/issues/27265
PROPOSAL: https://github.com/Expensify/App/issues/27265#issuecomment-1715990945

### Tests

Test 1

1. Create a workspace
2. Go to the workspace chat created
3. Click on the header
4. Click on 'Share Code' in right hand panel
5. Verify that the workspace name is shown below the QR code

Test 2

1. Go to an archived room
2. Click on the header
3. Click on 'Share Code' in right hand panel
4. Verify that the room name is shown as "roomName (archived)" below QR code

Test 3

1. Reply to any message in the workspace to create a thread
2. Go to the chat thread in a workspace chat
3. Click on the header
4. Click on 'Share Code' in right hand panel
5. Verify that `threadName` is shown as `title` and `workSpace` is shown as `subtitle` below QR code.

Test 4

1. Create a task in a workspace chat
2. Click on the task header
3. Click on 'Share Code' in right hand panel
4. Verify that `taskName` is shown as `title` and `workSpace` is shown as `subtitle` below QR code.

- [x] Verify that no errors appear in the JS console

### Offline tests

Test 1

1. Create a workspace
2. Go to the workspace chat created
3. Click on the header
4. Click on 'Share Code' in right hand panel
5. Verify that the workspace name is shown below the QR code

Test 2

1. Go to an archived room
2. Click on the header
3. Click on 'Share Code' in right hand panel
4. Verify that the room name is shown as "roomName (archived)" below QR code

Test 3

1. Reply to any message in the workspace to create a thread
2. Go to the chat thread in a workspace chat
3. Click on the header
4. Click on 'Share Code' in right hand panel
5. Verify that `threadName` is shown as `title` and `workSpace` is shown as `subtitle` below QR code.

Test 4

1. Create a task in a workspace chat
2. Click on the task header
3. Click on 'Share Code' in right hand panel
4. Verify that `taskName` is shown as `title` and `workSpace` is shown as `subtitle` below QR code.

### QA Steps

Test 1

1. Create a workspace
2. Go to the workspace chat created
3. Click on the header
4. Click on 'Share Code' in right hand panel
5. Verify that the workspace name is shown below the QR code

Test 2

1. Go to an archived room
2. Click on the header
3. Click on 'Share Code' in right hand panel
4. Verify that the room name is shown as "roomName (archived)" below QR code

Test 3

1. Reply to any message in the workspace to create a thread
2. Go to the chat thread in a workspace chat
3. Click on the header
4. Click on 'Share Code' in right hand panel
5. Verify that `threadName` is shown as `title` and `workSpace` is shown as `subtitle` below QR code.

Test 4

1. Create a task in a workspace chat
2. Click on the task header
3. Click on 'Share Code' in right hand panel
4. Verify that `taskName` is shown as `title` and `workSpace` is shown as `subtitle` below QR code.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>
Workspace Room
workspaceRoom

Owner Workspace Chat
ownerWorkspaceChat

Owner view of Chat of Invited Member to Workspace
ownerViewOfChatOfInvitedMemberToWorkspace

Archived Workspace Chat
archivedWorkspaceChat

Archived Room
archivedRoom

Task in Room
taskInRoom

Task in Workspace
taskinWorkspace-

Subtask inside Task in Workspace
subTaskInsideTaskInWorkspace

Thread in Workspace
threadInWorkspace

Thread in Room of Workspace
threadInRoomOfWorkspace

Non-workspace Thread
nonWorkspaceThread

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/102477862/c42e81b6-c6fa-4a12-a1a7-c9e50784a82f

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/102477862/2f064f11-a4f0-43b9-b457-5fb2a966f016

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/102477862/3d56262c-e071-4412-aa40-5327fe460c50

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/102477862/af366a44-6f2b-4a87-b0bc-f063356d7626

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/102477862/c5f91a01-4f53-434d-b10b-c2b644253763

</details>

### CODE REVIEW

**Review by 0xmiroslav — APPROVED** [review]
LGTM!
Asked for design confirmation [here](https://github.com/Expensify/App/issues/27265#issuecomment-1721220616)

**Comment by dangrous** [comment]
Code looks good! I commented on the issue but I'd like to wait for design to confirm that this works for them too. I made one suggestion on how to make it a little easier for them to confirm. Thanks!

**Comment by c3024** [comment]
@dangrous 
Shall I push a commit using `getReportName` for all report QR codes?

**Comment by dangrous** [comment]
Sounds like yes! We can just update this PR if that works for you.

Did you see the comment about the subtitle? Do you think you can handle that here as well?

**Comment by c3024** [comment]
I updated tests and screenshots for all cases that changed from existing behaviour. Please review and suggest any changes if required. @0xmiroslav 

In the context of this comment [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1695015289529269?thread_ts=1695004908.885709&cid=C01GTK53T8Q), I just want to clarify that for a thread or task created directly in a workspace chat, the QR code subtitle shows workspaceName like this as asked.
chatThread
But for a thread created inside a room (custom or # admins etc.) the subtitle in the QR code shows the roomName like this.
Screenshot 2023-09-18 at 5 25 15 PM
I think this is fine and what we expect. Please let me know if we need to change anything else. @dangrous

File: src/pages/ShareCodePage.js
Comment (dangrous) [inline src/pages/ShareCodePage.js:47]:
```suggestion
        const subtitle = isReport ? (ReportUtils.getChatRoomSubtitle(this.props.report) || ReportUtils.getParentNavigationSubtitle(this.props.report).rootReportName) : formattedEmail;
```
just to make this a little easier to understand

**Review by dangrous — CHANGES_REQUESTED** [review]
One tiny change - this looks good though, we can merge once this is done!

**Comment by dangrous** [comment]
Oh wait sorry - I just noticed. So in the Slack thread Jason suggested that the subtitle for the threads  (and tasks, IOUs, etc.) should be the workspace name (if it exists), rather than the parent thread. Can we make that change?

**Comment by c3024** [comment]
I pushed a commit. Why are the changes not updated here?

**Comment by c3024** [comment]
I think this commit settles all we want.
1. Threads and tasks inside rooms show subtitle as workspace and not the name of the room
2. I understand [this comment](https://expensify.slack.com/archives/C01GTK53T8Q/p1695014543196679?thread_ts=1695004908.885709&cid=C01GTK53T8Q) as
  - owner should have title of QR code as workspace, and subtitle as the text `Workspace`
  - owner's view for an invitied member chat in workspace shows the invited member as title, and workspace name as subtitle. (please refer screenshots for more clarity)
3. Non-workspace threads have no subtitle which is okay as per the Slack discussion.
@dangrous please suggest if I missed anything.
cc @0xmiroslav

File: src/pages/ShareCodePage.js
Comment (dangrous) [reply src/pages/ShareCodePage.js]: weird! never mind then haha

**Review by dangrous — APPROVED** [review]
Yep those screenshots look good - I think we're good to go!

**Comment by dangrous** [comment]
@0xmiroslav do you mind taking one last look as well? I'm not sure why it didn't dismiss your review since there have been changes.

**Review by 0xmiroslav — APPROVED** [review]
LGTM2
Not sure if it's bug but once C+ approved, ✅ keeps remaining even though author pushed more changes 😄

**Review by dangrous — COMMENTED** [review]
not sure why my review got dismissed, but I approved this!

**Pull request merged**
